Fix const type error in reverse mode

sudo-panda opened the pull request:
This PR fixes the error that arises due to const type parameters or variables being used in the input function. The const type was being carried forward when declaring clad tape, diff variables and temporary variables. Since these variables are modified later it gave an error during code generation.

vgvassilev reviewed (commented):
Nicely done! Could you add some tests for the new code?

vgvassilev commented:
Can you squash the test commit into the commit with the changes. Is the other code path that codecov complains about testable?

sudo-panda commented:
> Is the other code path that codecov complains about testable?

I tried to reach it using
```
int x;

int &g(int a) {
  x = a;
  return x;
}

double f(double a, double b) {
  g(b) /= a;
  return x;
}
```
and
```
double f(double a, double b) {
  volatile int r = 8;
  r /= a;
  return r;
}
```
but couldn't. Do you have any ideas?

vgvassilev commented:
> # [Codecov](https://codecov.io/gh/vgvassilev/clad/pull/236?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Vassil+Vassilev) Report
> 
> > Merging [#236](https://codecov.io/gh/vgvassilev/clad/pull/236?src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Vassil+Vassilev) ([681d2c0](https://github.com/vgvassilev/clad/commit/681d2c013cf2d02588a01202126a19eb473bf86f)) into [master](https://codecov.io/gh/vgvassilev/clad/commit/a9b7e2fab6a34dd6de298316031722584940c831?el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Vassil+Vassilev) ([a9b7e2f](https://github.com/vgvassilev/clad/commit/a9b7e2fab6a34dd6de298316031722584940c831)) will **increase** coverage by `0.05%`.
> > The diff coverage is `91.30%`.
> 
> > ❗ Current head [681d2c0](https://github.com/vgvassilev/clad/commit/681d2c013cf2d02588a01202126a19eb473bf86f) differs from pull request most recent head [450ca3f](https://github.com/vgvassilev/clad/commit/450ca3f9a4c0b7fb404f40157abf6f5bb1e11cdf). Consider uploading reports for the commit [450ca3f](https://github.com/vgvassilev/clad/commit/450ca3f9a4c0b7fb404f40157abf6f5bb1e11cdf) to get more accurate results
> > [![Impacted file tree graph](https://camo.githubusercontent.com/fbd1b904a3888f3bd01e886138e7966a2c764852b73341ebb6be8bbb0f05b9a9/68747470733a2f2f636f6465636f762e696f2f67682f766776617373696c65762f636c61642f70756c6c2f3233362f6772617068732f747265652e7376673f77696474683d363530266865696768743d313530267372633d707226746f6b656e3d3966365134656d3868452675746d5f6d656469756d3d726566657272616c2675746d5f736f757263653d6769746875622675746d5f636f6e74656e743d636f6d6d656e742675746d5f63616d706169676e3d70722b636f6d6d656e74732675746d5f7465726d3d56617373696c2b56617373696c6576)](https://codecov.io/gh/vgvassilev/clad/pull/236?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Vassil+Vassilev)
> 
> ```diff
> @@            Coverage Diff             @@
> ##           master     #236      +/-   ##
> ==========================================
> + Coverage   84.08%   84.13%   +0.05%     
> ==========================================
>   Files          21       21              
>   Lines        2689     2704      +15     
> ==========================================
> + Hits         2261     2275      +14     
> - Misses        428      429       +1     
> ```
> 
> [Impacted Files](https://codecov.io/gh/vgvassilev/clad/pull/236?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Vassil+Vassilev) 	Coverage Δ 	
> [lib/Differentiator/ReverseModeVisitor.cpp](https://codecov.io/gh/vgvassilev/clad/pull/236/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Vassil+Vassilev#diff-bGliL0RpZmZlcmVudGlhdG9yL1JldmVyc2VNb2RlVmlzaXRvci5jcHA=) 	`90.50% <83.33%> (-0.06%)` 	⬇️
> [include/clad/Differentiator/ReverseModeVisitor.h](https://codecov.io/gh/vgvassilev/clad/pull/236/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Vassil+Vassilev#diff-aW5jbHVkZS9jbGFkL0RpZmZlcmVudGlhdG9yL1JldmVyc2VNb2RlVmlzaXRvci5o) 	`98.36% <100.00%> (+0.32%)` 	⬆️
> [Impacted Files](https://codecov.io/gh/vgvassilev/clad/pull/236?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Vassil+Vassilev) 	Coverage Δ 	
> [lib/Differentiator/ReverseModeVisitor.cpp](https://codecov.io/gh/vgvassilev/clad/pull/236/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Vassil+Vassilev#diff-bGliL0RpZmZlcmVudGlhdG9yL1JldmVyc2VNb2RlVmlzaXRvci5jcHA=) 	`90.50% <83.33%> (-0.06%)` 	⬇️
> [include/clad/Differentiator/ReverseModeVisitor.h](https://codecov.io/gh/vgvassilev/clad/pull/236/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Vassil+Vassilev#diff-aW5jbHVkZS9jbGFkL0RpZmZlcmVudGlhdG9yL1JldmVyc2VNb2RlVmlzaXRvci5o) 	`98.36% <100.00%> (+0.32%)` 	⬆️

Could you add this diff in a separate commit in a try to fix codecov:

```diff
diff --git a/.github/workflows/ci.yml b/.github/workflows/ci.yml
index 69c9ceb..c05807b 100644
--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -233,12 +233,6 @@ jobs:
             compiler: gcc-5
             clang-runtime: '11'
 
-          - name: ubu18-gcc6-runtime11-analyzers
-            os: ubuntu-18.04
-            compiler: gcc-6
-            clang-runtime: '11'
-            coverage: true
-
           - name: ubu18-gcc6-runtime11-coverity
             os: ubuntu-18.04
             compiler: gcc-6
@@ -255,6 +249,12 @@ jobs:
             compiler: gcc-9
             clang-runtime: '11'
 
+          - name: ubu18-gcc6-runtime11-analyzers
+            os: ubuntu-18.04
+            compiler: gcc-9
+            clang-runtime: '11'
+            coverage: true
+
           - name: ubu16-clang4-runtime11
             os: ubuntu-16.04
             compiler: 'clang-4.0'
```

sudo-panda commented:
@vgvassilev Shouldn't the new name be `ubu18-gcc9-runtime11-analyzers`

vgvassilev commented:
> @vgvassilev Shouldn't the new name be `ubu18-gcc9-runtime11-analyzers`

Yeah, sorry. Good catch.

vgvassilev commented:
It looks like we should 'match' the version of gcov which seems to be 7.5 (see [here](https://github.com/vgvassilev/clad/pull/236/checks?check_run_id=2756415410#step:20:24)). Can you try using gcc 7? Or alternatively we should install the compiler-compatible version of gcov which seems more sustainable.

vgvassilev commented:
@sudo-panda, hm, the same issue again... Could you apply that patch:
```diff
diff --git a/.github/workflows/ci.yml b/.github/workflows/ci.yml
index c316859..b670de2 100644
--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -630,7 +630,7 @@ jobs:
 
     - name: Upload to codecov.io
       if: ${{ success() && (matrix.coverage == true) }}
-      uses: codecov/codecov-action@v1
+      uses: codecov/codecov-action@v1.0.5
       with:
         file: ./coverage.info
```

vgvassilev commented:
Maybe:
```diff
diff --git a/.github/workflows/ci.yml b/.github/workflows/ci.yml
index c316859..f75a0a2 100644
--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -630,9 +630,10 @@ jobs:
 
     - name: Upload to codecov.io
       if: ${{ success() && (matrix.coverage == true) }}
-      uses: codecov/codecov-action@v1
+      uses: codecov/codecov-action@v1.0.5
       with:
         file: ./coverage.info
+        verbose: true
 
     - name: Upload report to coverity_scan
       if: ${{ success() && (matrix.coverity == true) }}
```

vgvassilev commented:
Also looks like the newest version is `v1.5.0`. We can try that as well.

sudo-panda commented:
@vgvassilev have you checked this error:
https://github.com/vgvassilev/clad/pull/236/checks?check_run_id=2775111043#step:21:54

vgvassilev commented:
> @vgvassilev have you checked this error:
> https://github.com/vgvassilev/clad/pull/236/checks?check_run_id=2775111043#step:21:54

Yes, that's the one I used to have and fixed (I thought) in #239... Do you have something particular in the way you forked? It should be able to upload the information from forks as token is not required. I believe the codecov folks screwed up, as all of that worked few days ago...

sudo-panda commented:
> Yes, that's the one I used to have and fixed (I thought) in #239... Do you have something particular in the way you forked?

I used the github fork button to fork. That's the only way I know. If you are taking about creating a new branch. I made this branch from `master` and then cherry-picked the commit from my old branch `array-diff-support`. I rebased the branch on top of `coverage-fix` branch to see if the coverage part of it was fixed and once you merged your PR into `master` I rebased this back on `master`.

vgvassilev commented:
Nothing uncommon. Weird, will have to contact the codecov team.

vgvassilev commented:
> Also looks like the newest version is `v1.5.0`. We can try that as well.

@sudo-panda, can you bump to this, just to check?

sudo-panda commented:
> > Also looks like the newest version is `v1.5.0`. We can try that as well.
> 
> @sudo-panda, can you bump to this, just to check?

Was doing just that

vgvassilev commented:
Ok, another random thing to try:
```diff
diff --git a/.github/workflows/ci.yml b/.github/workflows/ci.yml
index c316859..c5e0a70 100644
--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -337,8 +337,6 @@ jobs:
 
     steps:
     - uses: actions/checkout@v2
-      with:
-        fetch-depth: 0
     - name: Set up Python
       uses: actions/setup-python@v2
       with:
```

vgvassilev commented:
@sudo-panda, can you remove the ci-related commits and rebase?
